sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -0,0 +1,362 @@
+            querySpec.where(new WhereClause(query));
+        }
+
+        if (querySpec.orderBy().isPresent()) {
+            OrderBy orderBy = querySpec.orderBy().get();
+            List<Symbol> orderBySymbols = FieldReplacingVisitor.INSTANCE.process(orderBy.orderBySymbols(), null);
+            querySpec.orderBy(new OrderBy(orderBySymbols, orderBy.reverseFlags(), orderBy.nullsFirst()));
+        }
+
+        if (querySpec.groupBy().isPresent()) {
+            List<Symbol> groupBy = FieldReplacingVisitor.INSTANCE.process(querySpec.groupBy().get(), null);
+            querySpec.groupBy(groupBy);
+        }
+
+        if (querySpec.having().isPresent() && !querySpec.having().get().noMatch()) {
+            Symbol query = FieldReplacingVisitor.INSTANCE.process(querySpec.having().get().query(), null);
+            querySpec.having(new HavingClause(query));
+        }
+    }
+
+    private boolean hasNestedAggregations(QueriedSelectRelation relation) {
+        QuerySpec querySpec1 = relation.querySpec();
+        QuerySpec querySpec2 = relation.relation().querySpec();
+
+        if ((querySpec1.hasAggregates() || querySpec1.groupBy().isPresent()) &&
+            (querySpec2.hasAggregates() || querySpec2.groupBy().isPresent() || querySpec2.orderBy().isPresent())) {
+            return true;
+        }
+
+        return querySpec1.where().hasQuery() && querySpec1.where() != WhereClause.MATCH_ALL &&
+               AggregateFunctionReferenceVisitor.any(querySpec1.where().query());
+    }
+
+    static class Context {
+        private final AnalysisMetaData analysisMetaData;
+        private final List<Field> fields;
+
+        private QuerySpec querySpec;
+
+        public Context(AnalysisMetaData analysisMetaData, List<Field> fields) {
+            this.analysisMetaData = analysisMetaData;
+            this.fields = fields;
+        }
+
+        public Collection<? extends Path> paths() {
+            return Collections2.transform(fields, new com.google.common.base.Function<Field, Path>() {
+                @Override
+                public Path apply(Field input) {
+                    return input.path();
+                }
+            });
+        }
+    }
+
+    private static class FieldReplacingVisitor extends ReplacingSymbolVisitor<Void> {
+
+        public static final FieldReplacingVisitor INSTANCE = new FieldReplacingVisitor(true);
+        private static final FieldRelationVisitor<Symbol> FIELD_RELATION_VISITOR = new FieldRelationVisitor<>(INSTANCE);
+
+        private FieldReplacingVisitor(boolean inPlace) {
+            super(inPlace);
+        }
+
+        @Override
+        public Symbol visitField(Field field, Void context) {
+            Symbol output = FIELD_RELATION_VISITOR.process(field.relation(), field);
+            return output != null ? output : field;
+        }
+    }
+
+    private static class AggregateFunctionReferenceVisitor extends SymbolVisitor<Void, Boolean> {
```

Comment:
Can the visitors here be renamed to better indicate what they're used for instead of how they internally operate?